Adding Plot ESS Evolution

imperorrp opened the pull request:
First commit for `plot_ess_evolution()`, as a branched off plot from legacy `plot_ess` for kind = 'evolution' (https://github.com/arviz-devs/arviz-plots/issues/5). The `line_xy` and `scatter_xy` visual elements were used for the plotting, with rug plot, axes labelling and title labelling functionality taken from ESS plot functionality. Since the ESS plot PR hasn't been merged yet, the `scatter_xy` visual element was duplicated for use here. 

#### WIP: 
Marking the 'bulk' and 'tail' artists with a legend 

#### Current plot outputs:

`azp.plot_ess_evolution(data)`
![image](https://github.com/user-attachments/assets/4c0b98df-7320-4f27-bfc6-bee7cb68fdb1)


`azp.plot_ess_evolution(data, var_names=["mu", "tau"])`
![image](https://github.com/user-attachments/assets/0e3fd641-e0db-473c-98cf-c9481f5d37df)


<!-- readthedocs-preview arviz-plots start -->
----
📚 Documentation preview 📚: https://arviz-plots--71.org.readthedocs.build/en/71/

<!-- readthedocs-preview arviz-plots end -->

OriolAbril commented on src/arviz_plots/plots/evolutionplot.py:
```diff
@@ -0,0 +1,449 @@
+        aes_map = aes_map.copy()
+    aes_map.setdefault("bulk", plot_collection.aes_set.difference({"overlay"}))
+    aes_map.setdefault("tail", plot_collection.aes_set.difference({"overlay"}))
+    aes_map.setdefault("mean", plot_collection.aes_set)
+    aes_map.setdefault("mean", ["linestyle", "-"])
+    aes_map.setdefault("sd", plot_collection.aes_set)
+    aes_map.setdefault("sd", ["linestyle", "-"])
```
I think by default we don't want any aesthetic mapping to be applied to either mean nor sd, so I would remove these 4 lines. Also note the lines 2 and 4 are not doing anything because the lines right before already define that key, but if they did, it would be incorrect. Values in the `aes_map` dict should be sets of strings that represent aesthetics.

This seems a confusion with setting a default linestyle in `plot_kwargs`. If you want an aesthetic property to always have the same value, use `plot_kwargs` if instead you want it to depend on a dimension/variable use `aes` in `pc_kwargs` to define the mapping and `aes_map` to indicate to which visual elements should the mapping be applied too.

There is nothing about this concepts in https://arviz-plots.readthedocs.io/en/latest/contributing/new_plot.html because that doc is for writing plots and these are usage questions which are therefore out of scope. The usage docs are instead https://arviz-plots.readthedocs.io/en/latest/tutorials/plots_intro.html and to a much lesser extent https://arviz-plots.readthedocs.io/en/latest/tutorials/intro_to_plotcollection.html

OriolAbril commented on src/arviz_plots/plots/evolutionplot.py:
```diff
@@ -0,0 +1,449 @@
+        :func:`~arviz_plots.visuals.line_xy`
+        * "tail" -> passed to :func:`~arviz_plots.visuals.scatter_xy` and
+        :func:`~arviz_plots.visuals.line_xy`
+        * divergence -> passed to :func:`~.visuals.trace_rug`
```
This key should be "rug" instead. The most common thing to use (and therefore default for `rug_kind` are divergences) but they are not the only option. As we agreed on smaller set of fixed keys in #66, I think this key should be "rug" instead of `rug_kind`.

Note this will require multiple changes in the function code

OriolAbril commented on src/arviz_plots/plots/evolutionplot.py:
```diff
@@ -0,0 +1,449 @@
+    n_draws = distribution.sizes["draw"]
+    n_samples = n_draws * distribution.sizes["chain"]
+    plot_bknd = import_module(f".backend.{backend}", package="arviz_plots")
+    default_bulk_color = plot_bknd.get_default_aes("color", 2, {})[0]
+    default_tail_color = plot_bknd.get_default_aes("color", 2, {})[1]
```
```suggestion
    default_bulk_color, default_tail_color = plot_bknd.get_default_aes("color", 2, {})
```

OriolAbril commented on src/arviz_plots/plots/evolutionplot.py:
```diff
@@ -0,0 +1,449 @@
+
+    # compute and add ess evolution subplots for 'bulk' and 'tail'
+
+    n_draws = distribution.sizes["draw"]
+    n_samples = n_draws * distribution.sizes["chain"]
```
This won't work when `sample_dims` are not `chain` and `draw`, it should work at least with a single `sample` dimension without users needing to set any extra parameter manually (more complex cases should also be possible but might require users to provide specific `pc_kwargs` or things like that).

OriolAbril commented on src/arviz_plots/plots/evolutionplot.py:
```diff
@@ -0,0 +1,449 @@
+
+        # ----------------------------
+
+        xdata = np.linspace(0, n_samples, n_points)
```
This is overwriting the xdata above but I think we want the one above, so I'd remove this line. Otherwise we get ess not necessarly 0 at 0 samples which makes no sense

OriolAbril commented on src/arviz_plots/plots/evolutionplot.py:
```diff
@@ -0,0 +1,449 @@
+            bulk_kwargs.setdefault("color", default_bulk_color)
+
+        # step 5
+        plot_collection.map(
+            scatter_xy, "bulk", data=ess_dataset, ignore_aes=bulk_ignore, **bulk_kwargs
+        )
+        plot_collection.map(
+            line_xy, "bulk", data=ess_dataset, ignore_aes=bulk_ignore, **bulk_kwargs
+        )
```
These are two different visual elements which should have different names and consequently different kwargs too. It should be possible to remove one but not the other, and it should be possible to change the marker of the scatter for example, but if we use the same kwargs, these are then passed to the line plotting function which doesn't accept the marker kwarg.

OriolAbril commented on src/arviz_plots/plots/evolutionplot.py:
```diff
@@ -0,0 +1,449 @@
+            pc_kwargs["aes"].setdefault("color", ["model"])
+            # setting x aesthetic to np.linspace(-x_diff/3, x_diff/3, length of 'model' dim)
+            # x_diff = span of x axis (1) divided by number of points to be plotted (n_points)
+            x_diff = 1 / n_points
+            if "x" not in pc_kwargs:
+                pc_kwargs["x"] = np.linspace(-x_diff / 3, x_diff / 3, distribution.sizes["model"])
```
I thought I was reviewing the plot_ess PR. Having x with an aesthetic mapping here makes no sense and should be completely removed.

imperorrp commented on src/arviz_plots/plots/evolutionplot.py:
```diff
@@ -0,0 +1,449 @@
+                divergence_kwargs.setdefault("color", "black")
+            if "marker" not in div_aes:
+                divergence_kwargs.setdefault("marker", "|")
+            # if "width" not in div_aes: # should this be hardcoded?
+            #    divergence_kwargs.setdefault("width", linewidth)
```
I tried this with bokeh, seems like there definitely is some default that isn't zero when nothing is used- 

![image](https://github.com/user-attachments/assets/bc67fda2-b8f7-48b7-9d38-490776f84d66)

OriolAbril replied:
then no need to set it, you can delete these lines

imperorrp commented on src/arviz_plots/plots/evolutionplot.py:
```diff
@@ -0,0 +1,449 @@
+        ylabel = ylabel.format("Relative ESS")
+    else:
+        ylabel = ylabel.format("ESS")
+    xlabel = "Total Number of Draws"
```
So "xlabel", "ylabel" as two separate keys, right?

OriolAbril replied:
yes

imperorrp commented:
Updated plot_ess_evolution. The points and lines for both 'bulk' and 'tail' methods can now be turned on and off via `plot_kwargs`, the mean, sd, and min_ess were added following the current plot_ess logic, and xlabel and ylabel texts can also now be set separately. 

Also added a common ess_dataset computing function since it can be called upto 4 times (one for each of the cartesian product of 'bulk', 'tail' and points, lines). 

The rug plot element is still WIP.

The outputs without the rug are now like: 
![image](https://github.com/user-attachments/assets/3d2fbcf2-01ad-4a7b-ba30-b61139cd9082)

![image](https://github.com/user-attachments/assets/24974d5d-759a-40ee-a19c-1f5fa317294e)

![image](https://github.com/user-attachments/assets/6d6f6138-d50b-414a-959e-64f3759d9d70)

#### An aesthetics issue with the color coding of the mean/sd plot elements: 

For the mean/sd plotting (and this'd similarly apply to plot_ess too) it looks like setting mean and sd to one color by default as I'm doing currently makes it hard to interpret that they have been overlapped when they coincide: 
![image](https://github.com/user-attachments/assets/0912c2cd-9b04-4b28-a9da-cb1811dbf492)

When they're not set to the same color, it's visible that they coincide and are overlapped but then the colors might be confusing as they correspond to the 'bulk' and 'tail' plotting elements' colors too:
![image](https://github.com/user-attachments/assets/26da5cdc-77b2-4566-a1a3-fb8722e82283)

OriolAbril commented:
I think annotating the lines is he only way to clearly indicate which is which. If you set `extra_methods=True` in current arviz, you'll see the output looks like this:

![imatge](https://github.com/user-attachments/assets/6a3eb83b-f6ea-433d-b1fd-83a490000e60)

I would add these two annotations as visual elements here too

imperorrp commented:
Rebased plot ess evolution commits

imperorrp commented:
The mean and sd lines are now annotated:

![image](https://github.com/user-attachments/assets/44ca883b-de7f-4b39-80fa-e2f4300fdb71)

Should the min_ess line also be annotated? (This also applies for ESS plot #58 ). Current Arviz ESS plot doesn't have it annotated (https://github.com/arviz-devs/arviz/blob/main/arviz/plots/backends/matplotlib/essplot.py#L150) but it would probably be nice to have and is easy to add.

imperorrp commented:
Fixed hypothesis tests errors and documentation build
